### ISSUE TITLE
[4.14][10.0][TAMA][AKARI] GCAM crashing

### ISSUE THREAD

**Issue by Gouster4** [sonyxperiadev/bug_tracker#544]
**Platform:Tama** 
**Device:Akari** 
**Kernel version:4.14** 
**Android version:10** 
**Software binaries version:v3** 

**Previously working on**
4.9 kernel

**Description**
GCAM doesnt work with 4.14 kernel

**Symptoms**
Google Camera app crashes shortly after opened.

**How to reproduce**
Install any google camera port.

**Additional context**
After GCAM crash, camera doesnt work with any other app until next reboot.
[logcat](http://techax.sk/gcam.log)

**Comment by ix5** [comment]
Could you please narrow this down? Maybe GCAM is using the v1 HAL?
In OpenCamera, there should be an option to switch between v1 and v3, please give that a try and supply logs.

**Comment by MartinX3** [comment]
[gcam_crash.log](https://github.com/sonyxperiadev/bug_tracker/files/4300406/gcam_crash.log)
I replicated the bug and shorted the log.

I searched for obvious messages and found:
```
03-06 22:01:51.350   760  2276 E CameraService: onTorchStatusChangedLocked: cannot get torch status of camera 1: No such file or directory (-2)
03-06 22:01:51.532   760  2276 E Camera3-Device: Camera 0: configureStreamsLocked: Unable to configure streams with HAL: Function not implemented (-38)
03-06 22:01:51.738   760  2276 E CameraDeviceClient: endConfigure: Camera 0: Error configuring streams: Function not implemented (-38)
03-06 22:01:51.740  9028  9106 E CameraCaptureSession: Session 0: Failed to create capture session; configuration failed
03-06 22:01:51.742  9028  9106 W CAM_SessionOpener: mmh: android.hardware.camera2.CameraAccessException: CAMERA_ERROR (3): endConfigure:500: Camera 0: Error configuring streams: Function not implemented (-38)
03-06 22:01:52.086  9028  9053 E PhenotypeFlagCommitter: Retrieving snapshot for com.google.android.apps.camera#com.google.android.GoogleCamera failed
03-06 22:01:52.086  9028  9053 E PhenotypeFlagCommitter: java.util.concurrent.ExecutionException: java.lang.SecurityException: GoogleCertificatesRslt: not whitelisted: pkg=com.google.android.GoogleCamera, sha1=d32102739a5ef8d9fef2faac2e3780781ea8a082, atk=true, ver=200616039.true (go/gsrlt)
03-06 22:01:54.807  9028  9152 E CAM_DirectGyro: Unable to create gyro direct channel
03-06 22:01:54.807  9028  9152 E CAM_DirectGyro: java.io.UncheckedIOException: java.io.IOException: create HardwareBuffer direct channel failed -12
```

I hope that helps.

**Comment by mhtvsSFrpHdE** [comment]
Check if you have installed `Google Play Service` or `MicroG Services Core`
https://microg.org/download.html

The GCam not just require RAW capture, it will also require  
any implementation of Google Play Service.

I can run GCam and take photo using Xperia XZ Premium,  
but before HDR+ render process complete the phone will freeze and reboot.

I currently suspect that this rendering process overheats the phone  
The 835 big CPU clusters can hot very fast in seconds.

If I can have root on SONY AOSP, I will try to manual set CPU to run at lower clock speed  
but didn't find a way to root it right now.

**Comment by MartinX3** [comment]
@mhtvsSFrpHdE only the tama platform has the gcam problem.
It's a different problem.

**Comment by mhtvsSFrpHdE** [comment]
> 
> 
> @mhtvsSFrpHdE only the tama platform has the gcam problem.
> It's a different problem.
> 
> Please open a different issue ticket.

@MartinX3 I read again and realize other apps also crash not just GCAM  
so that's indeed different

**Comment by thiscantbeserious** [comment]
I can confirm that this also happens for Pioneer - this was tested on official PixelExperience Rom with V9C (unrelated to microG or anything). 

The camera sometimes works but then begins to lag and/or gets really dark I have to restart the Phone because any GCam (that previously) worked just close after trying to opening them.

So exactly like reported by OP, just with Pioneer

**Comment by Gouster4** [comment]
I think its another issue as on akari i tryed ton of gcam versions and every one crashes at startup, never got any gcam working, not even for a while since akari migrated to 4.14 kernel...

**Comment by sraase** [comment]
Nile uses the mm-camera stack, Tama uses CamX. So this is a different issue.
The main issue in the provided log is this:

> CHIUSECASE: [ERROR  ] chxusecaseutils.cpp:867 DefaultMatchingUsecase() Failed to match usecase. pSelectedUsecase is NULL
> CHIUSECASE: [ERROR  ] chxextensionmodule.cpp:1847 InitializeOverrideSession() For cameraId = 0 CreateUsecaseObject failed
> Camera3-Device: Camera 0: configureStreamsLocked: Unable to configure streams with HAL: Function not implemented (-38)

So gcam triggers an unsupported usecase.

**Comment by MartinX3** [comment]
@sraase is there any chance to fix this ? :)
The gcam apps are good for post processing on tama. :)

**Comment by sraase** [comment]
@MartinX3 I don't have a Tama device for testing available, so I don't know the type of unsupported configuration gcam triggers. When working on bring-up, I mostly use smaller test applications (think CameraTest) or OpenCamera. So I need more logs to be able to look into this, and I can't promise anything.

**Comment by MartinX3** [comment]
@sraase Oh, that's sad to hear.

Which kind of logs do you need anyone of us could provide? :)

**Comment by sraase** [comment]
I definitely need the requested stream configuration (i.e. how many streams are requested; their sizes, formats and maybe gralloc usage flags - that is the HAL configureStreams call) which trigger the unsupported usecase. That should allow me to reproduce the failure with a test application.

**Comment by MartinX3** [comment]
@sraase if a logcat (https://github.com/sonyxperiadev/bug_tracker/issues/544#issuecomment-595967248) isn't enough, could you give me an advice how to gather the correct logs? :)
I'm sadly not experienced in camera development. :(

**Comment by MarijnS95** [comment]
@MartinX3 It isn't hard to grep `logcat` for `configure_streams`:

```
02-12 10:57:42.163   654  2272 I CamX    : [CONFIG][HAL    ] camxhal3.cpp:904 configure_streams() HalOp: Begin CONFIG: 0x7e1cb8f410, logicalCameraId: 0, cameraId: 0
02-12 10:57:42.403   654  2272 I CamX    : [CONFIG][HAL    ] camxhal3.cpp:980 configure_streams()  FINAL stream[0] = 0x7e22430528 - info:
02-12 10:57:42.404   654  2272 I CamX    : [CONFIG][HAL    ] camxhal3.cpp:983 configure_streams()             format       : 34, HAL_PIXEL_FORMAT_IMPLEMENTATION_DEFINED
02-12 10:57:42.404   654  2272 I CamX    : [CONFIG][HAL    ] camxhal3.cpp:985 configure_streams()             width        : 1920
02-12 10:57:42.404   654  2272 I CamX    : [CONFIG][HAL    ] camxhal3.cpp:987 configure_streams()             height       : 1080
02-12 10:57:42.404   654  2272 I CamX    : [CONFIG][HAL    ] camxhal3.cpp:990 configure_streams()             stream_type  : 00000000, CAMERA3_STREAM_OUTPUT
02-12 10:57:42.405   654  2272 I CamX    : [CONFIG][HAL    ] camxhal3.cpp:992 configure_streams()             usage        : 10020100
02-12 10:57:42.405   654  2272 I CamX    : [CONFIG][HAL    ] camxhal3.cpp:994 configure_streams()             max_buffers  : 8
02-12 10:57:42.405   654  2272 I CamX    : [CONFIG][HAL    ] camxhal3.cpp:997 configure_streams()             rotation     : 00000000, CAMERA3_STREAM_ROTATION_0
02-12 10:57:42.406   654  2272 I CamX    : [CONFIG][HAL    ] camxhal3.cpp:1000 configure_streams()             data_space   : 00000000, HAL_DATASPACE_UNKNOWN
02-12 10:57:42.406   654  2272 I CamX    : [CONFIG][HAL    ] camxhal3.cpp:1002 configure_streams()             priv         : 0x7e30d52380
02-12 10:57:42.406   654  2272 I CamX    : [CONFIG][HAL    ] camxhal3.cpp:1004 configure_streams()             reserved[0]         : 0x7e30d52388
02-12 10:57:42.406   654  2272 I CamX    : [CONFIG][HAL    ] camxhal3.cpp:1006 configure_streams()             reserved[1]         : 0x0
02-12 10:57:42.407   654  2272 I CamX    : [CONFIG][HAL    ] camxhal3.cpp:1020 configure_streams()    pHalStream: 0x7e30d52388 format : 0x7fa30c06, overrideFormat : 0x7fa30c06 consumer usage: 0, producer usage: 10020100
02-12 10:57:42.407   654  2272 I CamX    : [CONFIG][HAL    ] camxhal3.cpp:980 configure_streams()  FINAL stream[1] = 0x7e17922b28 - info:
02-12 10:57:42.407   654  2272 I CamX    : [CONFIG][HAL    ] camxhal3.cpp:983 configure_streams()             format       : 33, HAL_PIXEL_FORMAT_BLOB
02-12 10:57:42.408   654  2272 I CamX    : [CONFIG][HAL    ] camxhal3.cpp:985 configure_streams()             width        : 5312
02-12 10:57:42.408   654  2272 I CamX    : [CONFIG][HAL    ] camxhal3.cpp:987 configure_streams()             height       : 2988
02-12 10:57:42.408   654  2272 I CamX    : [CONFIG][HAL    ] camxhal3.cpp:990 configure_streams()             stream_type  : 00000000, CAMERA3_STREAM_OUTPUT
02-12 10:57:42.408   654  2272 I CamX    : [CONFIG][HAL    ] camxhal3.cpp:992 configure_streams()             usage        : 00020003
02-12 10:57:42.409   654  2272 I CamX    : [CONFIG][HAL    ] camxhal3.cpp:994 configure_streams()             max_buffers  : 8
02-12 10:57:42.409   654  2272 I CamX    : [CONFIG][HAL    ] camxhal3.cpp:997 configure_streams()             rotation     : 00000000, CAMERA3_STREAM_ROTATION_0
02-12 10:57:42.409   654  2272 I CamX    : [CONFIG][HAL    ] camxhal3.cpp:1000 configure_streams()             data_space   : 08c20000, HAL_DATASPACE_V0_JFIF
02-12 10:57:42.410   654  2272 I CamX    : [CONFIG][HAL    ] camxhal3.cpp:1002 configure_streams()             priv         : 0x7e1d246a80
02-12 10:57:42.410   654  2272 I CamX    : [CONFIG][HAL    ] camxhal3.cpp:1004 configure_streams()             reserved[0]         : 0x7e1d246a88
02-12 10:57:42.410   654  2272 I CamX    : [CONFIG][HAL    ] camxhal3.cpp:1006 configure_streams()             reserved[1]         : 0x0
02-12 10:57:42.410   654  2272 I CamX    : [CONFIG][HAL    ] camxhal3.cpp:1020 configure_streams()    pHalStream: 0x7e1d246a88 format : 0x21, overrideFormat : 0x21 consumer usage: 0, producer usage: 20003
02-12 10:57:42.411   654  2272 I CamX    : [CONFIG][HAL    ] camxhal3.cpp:1026 configure_streams() HalOp: End CONFIG, logicalCameraId: 0, cameraId: 0
```

**Comment by sraase** [comment]
@MartinX3 Did you have time to grab a full crash report? I assume that if gcam fails on Tama, it will also fail on all other platforms using CamX (but I haven't verified).

**Comment by MartinX3** [comment]
@sraase  Sadly it's busy here with exams and my job. I even had not time to move back to a sodp based ROM. But it is on my todo list :)

As far as I know tama is the only camX platform with this problem, which confuses me.

**Comment by sraase** [comment]
> As far as I know tama is the only camX platform with this problem, which confuses me.

That is an important information. Can someone provide the GCAM output on a Kumano device (i.e. configureStreams and CHI-related logcat)? It would be very useful to know which topology is triggered by GCAM and compare between sdm845/sm8150. The configureStreams call for Tama should be the same.

**Comment by sraase** [comment]
> As far as I know tama is the only camX platform with this problem, which confuses me.
This is an important information. However, I cannot reproduce the issue on Android 11. (I could not get GCAM to run either, but there is a good chance that I am doing something wrong.)

@MartinX3 When you have time again, please try to confirm whether the issue still persists.

**Comment by sraase** [comment]
@MartinX3 I sent you an e-mail with more detailed instructions. Can you confirm that you received it?

**Comment by MartinX3** [comment]
Hey @sraase thank you for the helpful instructions and I got it! :)
Now after I finished the last bits of work before starting to study for my exams I will compile a new AOSP 11 image and migrate back to SODP again and try to gather the logs. :)

**Comment by MartinX3** [comment]
@sraase I followed everything you told me in your email (after half a day of trouble getting aosp + gapps to work, in the end only deactivating vbmeta made the gapps appear on my device...)
Here is the log
[camera.log](https://github.com/sonyxperiadev/bug_tracker/files/6049307/camera.log)

**Comment by MarijnS95** [comment]
Not sure what kind of debugging was enabled nor why that has to be handled in some non-transparent way through email, but here is GCAM requesting at least one incompatible usecase:

```
02-26 12:28:58.378  9059  9158 I CamX    : [CONFIG][HAL    ] camxhal3.cpp:904 configure_streams() HalOp: Begin CONFIG: 0x713561ace0, logicalCameraId: 0, cameraId: 0
02-26 12:28:58.379  9059  9158 E CHIUSECASE: [CONFIG ] chxsensorselectmode.cpp:587 FindBestSensorMode() Selected Usecase: 7, SelectedMode W=5504, H=3104, FPS:30, NumBatchedFrames:0, modeIndex:0
02-26 12:28:58.379  9089  9089 W CAM_Log : Tag ZoomLockViewController is 3 chars longer than limit.
02-26 12:28:58.379  9059  9158 E CHIUSECASE: [CONFIG ] chxextensionmodule.cpp:1804 InitializeOverrideSession() Session_parameters FPS range 8:30, BatchSize: 1 FPS: 30 SkipPattern: 1, cameraId = 0 selected use case = 1
02-26 12:28:58.380  9089  9089 W CAM_Log : Tag ShutterButtonAnimator is 2 chars longer than limit.
02-26 12:28:58.380  2916  2930 I ndroid.systemu: NativeAlloc concurrent copying GC freed 1795(180KB) AllocSpace objects, 0(0B) LOS objects, 49% free, 8127KB/15MB, paused 31us total 109.351ms
02-26 12:28:58.380  9059  9158 E CHIUSECASE: [ERROR  ] chxusecaseutils.cpp:867 DefaultMatchingUsecase() Failed to match usecase. pSelectedUsecase is NULL
02-26 12:28:58.381  9059  9158 E CHIUSECASE: [ERROR  ] chxextensionmodule.cpp:1847 InitializeOverrideSession() For cameraId = 0 CreateUsecaseObject failed
02-26 12:28:58.381  9059  9158 E CamX    : [ERROR][HAL    ] camxhaldevice.cpp:918 ConfigureStreams() CHI Module failed to configure streams
02-26 12:28:58.381  9059  9158 I CamX    : [CONFIG][HAL    ] camxhal3.cpp:1026 configure_streams() HalOp: End CONFIG, logicalCameraId: 0, cameraId: 0
02-26 12:28:58.381  9058  9064 E Camera3-Device: Camera 0: configureStreamsLocked: Unable to configure streams with HAL: Function not implemented (-38)
```

**Comment by sraase** [comment]
I apologize for the intransparency.

Anyway, after lots of wrestling, I have been finally able to reproduce this problem. GCAM creates four streams (IMPL_DEF @ 1440x1080, YUV @ 1024x768, YUV @ 4000x3000, RAW @ sensor size). On Kumano, this configuration triggers a specific pipeline setup, but fails to do so on Tama. One probable cause is Tama's higher-resolution sensor.

That being said, I can't get GCAM to work on Kumano either. If GCAM can be used on Kumano, I do not know how.

**Comment by MarijnS95** [comment]
@sraase Someone told me that GCAM `MGC_8.1.101_A9_V0b` works "okay ish" on Kumano. Probably hit-and-miss, all these various "mods" are infinitely confusing and horrible.

**Comment by MartinX3** [comment]
@sraase 
Oh, that means that it's impossible to get gcam support on camx? :(
It was working on kernel 4.9 with android 9 OEM and helped much to work around the camera limitations.

**Comment by sraase** [comment]
@MartinX3 It will work; the fix should be part of the next binary release (com.qti.chi.override.so).

**Comment by jevgenp** [comment]
AOSP A11 / OEMv5 / TAMA / XZ2C
Still no luck with google camera :)
Tried these
![Screenshot_20210528-093430](https://user-images.githubusercontent.com/66970854/119940683-39c3ce80-bf98-11eb-9d90-3ed0e216ed1b.png)
All of them not even starting, except 6.2.030 (that worked on AOSP A9).
This one runs, but without buffer fixed enabled in settings - viewfinder freezes.
And no reaction on capture button.
![screen4](https://user-images.githubusercontent.com/66970854/119941085-ca9aaa00-bf98-11eb-8964-a7f078a38154.png)
[gcam.txt](https://github.com/sonyxperiadev/bug_tracker/files/6558824/gcam.txt)
Sorry in advance if this post created badly, noob :)

**Comment by MartinX3** [comment]
The issue `java.lang.IllegalArgumentException: supportsCameraApi:2340: Unknown camera ID 2` is new
@sraase @jerpelea 
Could you please reopen this issue ticket? :)

**Comment by MartinX3** [comment]
Done -> https://github.com/sonyxperiadev/bug_tracker/issues/695
